File: Solutions/C#/0035_SearchInsertPosition/SearchInsertPosition.cs
using System;

public class Solution
{
    public int SearchInsert(int[] nums, int target)
    {
        int left = 0, right = nums.Length - 1, mid;

        while (left < right)
        {
            mid = (left + right) / 2;
            if (nums[mid] == target)
            {
                return mid;
            }

            if (nums[mid] < target)
            {
                left = mid + 1;
            }
            else
            {
                right = mid - 1;
            }
        }

        if (nums[left] == target)
        {
            return left;
        }

        if (nums[left] > target)
        {
            return left;
        }

        return left + 1;
    }
}

File: Solutions/C++/Arrays/SearchInsertPosition.cpp
#include <string>
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int searchInsert(vector<int>& nums, int target) {
        int left = 0, right = nums.size() - 1;
        if(nums[left] > target)
            return 0;

        if(nums[right] < target)
            return right + 1;

        while(left <= right) {
            int middle = left + (right - left) / 2;
            if(nums[middle] == target)
                return middle;

            if(nums[middle] < target) {
                if(nums[middle + 1] > target)
                    return middle + 1;

                left = middle + 1;
            }
            else {
                if(nums[middle - 1] < target)
                    return middle;
                right = middle - 1;
            }
        }

        return -1;
    }
};